classification metrics

In [1]:
from sklearn import metrics

In [2]:
# c , h , f are classess corresponding to the 9 players 
y_actual  = ["c" , "h", "f", "c","h" , "f" , "c" ,  "h" , "h"] # actual classes corresponding 
# 9 players 
y_pred =   ["c" , "c", "f", "h","h"  ,"c" , "c" , "f" , "f"] # predicted class for each player 

In [5]:
cm = metrics.confusion_matrix(y_actual , y_pred, labels= ["c" , "h", "f"]) # ordering of 
#y_pred and y_actual is important in the arguement of funciton
print(cm) # confusion matrix for a multiclass classification problem
# interpretation of the confusion matrix 

[[2 1 0]
 [1 1 2]
 [1 0 1]]


In [6]:
# the rows represent actual values and the columns represent predicted values .as unlike binary classification
# we don't only have positve and negative classes but we have multiple classes.
# so the first row is for the class cricket and the second & third rows are for hockey and football class resp.

the first row says that :
we have 2 + 1 +0   = 3  samples for the cricket class , 2:TC(True cricket) 1:FH(false hockey) 0:Ff(false football) & so on for other rows....

In [7]:
print(metrics.classification_report(y_actual ,y_pred , labels = ["c" , "h" , "f"]))

              precision    recall  f1-score   support

           c       0.50      0.67      0.57         3
           h       0.50      0.25      0.33         4
           f       0.33      0.50      0.40         2

    accuracy                           0.44         9
   macro avg       0.44      0.47      0.43         9
weighted avg       0.46      0.44      0.43         9



- precision for class cricket  = TC / (TC +  FC(belonging to hockey)   +  FC(belonging to football) )
- recall for class cricket  = TC / TC + FH(belong to class cricket) + Ff(belongint to class cricket)
- precision cricket = 2  /  (2  + 1 +  1 ) = 2 / 4 = 0.5
- recall for cricket = 2 /  ( 2 + 1 + 0 ) = 2 /3  = 0.67

- Precision for hockey  class = TH / TH +  FH(belonging to cricket)  +   FH(belonging to football)
- recall for hockey class  = TH / (TH  + FC(belonging to hockey) +   Ff(belonging to hockey))
- precisoin hockey  = 1 / ( 1 + 1 + 0)  = 0.5 
- recall   hockey = 1 / ( 1 + 1 + 2) =  0.25

-  macro avg :is the arithmatic mean of metrics so, (0.50 + 0.50 + 0.33) /3 = 0.44 <-- precision macro avg
-  support is the no.of samples from each classess.
-  weighted average : weighted avg of metrics so ,(3\*0.50  + 4\*0.50  + 2\*0.33) / (3 + 4 + 2) = 0.46 